pbjson-build/src/generator/message.rs

```diff
@@ -486,14 +495,34 @@ fn write_deserialize_message<W: Write>(
             }
         }
 
+        if skip_unknown_fields {
+            writeln!(
+                writer,
+                "{}GeneratedField::__SkipField__ => {{",
+                Indent(indent + 4),
+            )?;
+            writeln!(
+                writer,
+                "{}let _ = map.next_value::<serde::de::IgnoredAny>()?;",
+                Indent(indent + 5),
+            )?;
+            writeln!(writer, "{}}}", Indent(indent + 4))?;
+        }
+
         writeln!(writer, "{}}}", Indent(indent + 3))?;
         writeln!(writer, "{}}}", Indent(indent + 2))?;
     } else {
         writeln!(
             writer,
-            "{}while map.next_key::<GeneratedField>()?.is_some() {{}}",
+            "{}while map.next_key::<GeneratedField>()?.is_some() {{",
             Indent(indent + 2)
         )?;
+        writeln!(
+            writer,
+            "{}let _ = map.next_value::<serde::de::IgnoredAny>()?;",
```

Comment:
Do you have a link to the docs that describe why this is necessary, I can't see anything in https://docs.serde.rs/serde/de/trait.MapAccess.html. Basically I'm wondering if this is an existing bug I should file a ticket for, and associate this PR as fixing :smile:

Reply:
I'm not 100% sure on the explanation, but what I understand is that that you have to consume all the values so the deserializer can update the position at which to parse/consume the next token/value, and this is a way for telling serde's deserializer to consume and ignore the value. If nothing is done, the key will be consumed, but the value won't, so the position of the deserializer would be at the value's start. The next thing the `MapAccess` deserializer will try to do is to parse a key but the next token won't look like a key (because it will be the value), and then you will have a parsing error because it was expecting `:` but found `,`. So this is not a bug, this is a pretty standard tool for building parsers to explicitly skip/drop tokens.

The docs don't make it easy to find and understand, but if you look at the [`IgnoredAny` docs](https://docs.serde.rs/serde/de/struct.IgnoredAny.html) you get more details, although the example uses `SeqAccess`, the idea is the same for `MapAccess`.

Reply:
Aah thanks for the explanation, so previously if you tried to serialize a non-empty JSON payload with an empty message, the deserializer would get out of sync :+1:

It occurs to me that perhaps we should change this to be consistent and error if fields are provided for an empty message, unless skip_unknown_fields is specified :thinking:

Reply:
If the flag is not set, it does fails trying to deserialize an empty message. There's also already a [test for this](https://github.com/influxdata/pbjson/blob/main/pbjson-test/src/lib.rs#L58-L62), so the behavior is preserved after this change.

Reply:
I guess the serde_json deserializer must be permissive in this respect